[1단계 - 보드 초기화, 기물 이동] 노랑(노주희) 미션 제출합니다.

yesjuhee opened the pull request:
안녕하세용 노랑입니다!
리뷰 잘 부탁드립니다 🙇‍♀️ 🙇 

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)
- [ ] 
### 선택한 점수의 이유를 적어주세요.
객체지향적인 코드를 짜기 위해 일급 컬렉션을 사용하거나 게터를 최대한 사용하지 않도록 노력했습니다.
다만 로직을 짜는 과정에서 충분한 고민을 하지 않아서 메서드 인덴트가 지켜지지 않은 부분이 몇가지 있습니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->

질문은 코드에 남기겠습니다!

theminjunchoi committed:
docs: 기능 요구 사항 작성

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
feat: 메뉴에 없는 입력을 했을 경우, 재입력을 받는다.

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
feat: 상 차림에 맞는 보드를 생성하여 출력한다.

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
refactor: depth 조정

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
chore: 기물표기법 안내문구 추가

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
refactor: 메서드 분리

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
feat: 병 이동 구현

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
feat: 마 이동 구현

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
feat: 상 이동 구현

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
refactor: 생성자 private 변경

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
feat: move 명령 구현

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
feat: 차 이동 구현

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
feat: 포 이동 구현

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
refactor: 메서드명 변경 및 indent 수정

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
refactor: 불필요한 계층 제거

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
refactor: 기물이 스스로 탸입을 판단하도록 변경

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
refactor: 패키지 분리

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
feat: 재입력 로직 구현

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
refactor: 재입력 로직 수정

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
docs: 기능 요구 사항 및 설명 추가

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
refactor: 일급컬렉션 MovingRules 객체 사용

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
refactor: Position을 레코드에서 클래스로 수정

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
refactor: 불필요한 메서드 제거

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

theminjunchoi committed:
refactor: getBoard 메서드 제거

Co-authored-by: yesjuhee <juhee0924@g.skku.edu>

yesjuhee committed:
refactor: 상수 위치 변경, final 키워드 추가

yesjuhee committed:
refactor: 메서드를 적절한 위치로 이동

yesjuhee commented on src/main/java/janggi/board/BoardGenerator.java:
```diff
@@ -0,0 +1,107 @@
+            SetupOption.LEFT_SETUP, BoardGenerator::generateLeftSetup
+    );
+
+    public static Board generate(SetupOption setupOption) {
+        return SETUP_MENU.get(setupOption).get();
+    }
```
### BoardGenerator 테스트?

이 메서드를 **테스트하는게 좋을지, 또한 테스트한다면 어떻게 테스트하는게 좋을지**에 대해 제이의 의견을 들어보고 싶습니다!
저의 생각을 정리해보자면 아래와 같습니다.

`BoardGenerator::generate(SetupOption setupOption)`은 상차림에 대한 정보를 받아서 거기에 해당하는 초기 상태의 보드를 만들어서 생성하는 메서드입니다. 우선 이 메서드가 하는 책임이 꽤 중요하기 때문에 `generate()`에 대한 테스트가 필요하다라고 생각을 했습니다. 그러면서 또 한편으로는 이 메서드들은 로직이라고 할 만한 것이 전혀 없고, 그냥 정해진 위치에 정해진 기물들을 생성하기만 하는데 이걸 테스트할 필요가 있나라는 생각이 들었습니다.

또한 테스트를 하기로 결심했을 때, 어떻게 테스트를 하는 것이 적절할지 잘 떠오르지 않았습니다. 먼저 `generate()` 메서드를 처음 만든 시점에는 `Board`의 API가 존재하지 않았기 때문에 게터나 리플렉션이 아니면 테스트할 수가 없습니다. 코드가 완성된 지금 다시 생각해 봤을 때는 `Board::getPiece()`나 `Board::isExistPiece()`와 같은 메서드를 사용해서 어떤 위치에 어떤 기물이 있는지 확인할 수 있을 것 같습니다. 만약 이런 방식으로 테스트를 작성한다고 하면 32개의 모든 말이 어떤 위치에 있는지 테스트를 하는 메서드가 될 것 같은데 이런 테스트는 잘 손이 가지 않을 것 같기도 합니다.

위와 같은 이유로 잘 답이 서지 않아서 제이의 생각을 들어보고 싶습니다!

JunHoPark93 replied:
게임의 시작을 테스트하는건데 필요하지 않을까요? 정해진 대로 기물이 초기화되는지 정의해주신 display메서드를 호출해서 print되는 문자열들을 검사하는정도로 해도 괜찮을것같습니다. 이렇게하면 32개의 말을 위치를 일일이 검사하지않아도되지않을까요?

개인적으로 일전에 리뷰할때 보드에 각 팀마다 2개씩만 있어야하는 기물인데 3개씩 있어도 게임이 정상 진행되는 경우를 본적이 있어서 테스트는 필요해보입니다.

yesjuhee replied:
오호 그렇군요
그런데 출력되는 문자열을 검사하는 테스트를 작성할 경우 도메인의 테스트가 뷰에 의존하게 되는 문제가 있을 것 같아요
장기 같은 경우는 출력 형식이 바뀌기 쉬울 것 같고, 결국 변화에 너무 취약한 테스트가 될 것 같아서 쉽게 손이 가지 않는데 이 부분은 어떻게 생각하시나욥?

JunHoPark93 replied:
어차피 toString 결과로 나올거니 괜찮지 않을까요?
말씀하신 단점도 존재는 하지만 너무 많은 경우의 수를 테스트하는것의 대안으로 제일 적절해보여서 제안드렸습니다. 
출력형식같은 경우는 바뀔수는 있겠지만 엄청 다양해지기는 힘들거같기도하구요.

yesjuhee commented on src/main/java/janggi/piece/Type.java:
```diff
@@ -0,0 +1,12 @@
+package janggi.piece;
+
+public enum Type {
+
+    CANNON,
+    CHARIOT,
+    ELEPHANT,
+    GENERAL,
+    GUARD,
+    HORSE,
+    SOLDIER
+}
```
### Piece가 어떤 기물에 해당하는지 구분하는 방법

포가 이동할 때 경로에 있는 기물이 포에 해당하는지 확인하거나, 기물들을 출력을 할 때 그 기물이 어떤 클래스인지 확인이 필요합니다.
저와 페어가 논의했을 때 가능한 방법은 아래와 같았습니다.
1. `getClass()` 사용
2. `instanceof` 사용
3. 보드의 구분자를 클래스 상수로 정의하고(`private static final`), 게터로 꺼내와서 구분
4. 보드의 구분자를 멤버변수로 정의하고(`private final`), 생성할 때 구분자를 설정하고, 게터로 꺼내와서 구분
5. `Type` enum을 정의하고 `Piece` 추상클래스에 `abstract Type getPiece()` 추상 메서드를 정의하고, 각 피스가 자신의 타입을 리턴하도록 사용(상수로 가지진 않음)

1,2번은 객체지향의 개념에 반하기 때문에 사용하지 않았고, 3~5 중에 구분하다가 최종적으로 5번으로 결정했습니다.
4,5는 거의 비슷한 것 같아서 3과 5를 비교하자면
5는 Type의 변화가 있을 때 `enum`을 수정해야한다는 점에서 OCP가 잘 지켜지지 않습니다.
3은 유연성이 높아지지만 상황을 통제하기 어려워지고 컨트롤해야 할 상태가 하나 늘어난다고 생각합니다.

리뷰어님이 보시기에 어떤 방법이 좋은 방법이라고 생각하시는지 궁금합니다!

JunHoPark93 replied:
저는 5번을 선택했을거같아요. 개인적으로 원칙도 중요하지만 최대한 간결하고 보기쉬운 코드를 선호합니다. 

첨언하면 저는 유연성을 가장하여 복잡성이 높아지는것은 경계하고있습니다. 내가 짠 코드라도 조금이라도 복잡하면 시간이 지나고 봤을때 이해가 안되는경우도 있어서요. 개인 의견이니 참고만해주세요~

그리고 type의 변화가 있는 경우가 자주있을까요?

yesjuhee replied:
저도 동의합니다ㅎㅎ
이 이상의 유연성을 고려하는 것은 경계해야할 것 같아요!

JunHoPark93 commented on README.md:
```diff
@@ -1,3 +1,86 @@
-장기 미션 저장소
+## 입출력 예시
+
+### 기물 표기법
+
+|   | 한 | 초 |
+|---|---|---|
```
readme 깔끔하게 잘 정리하셨네요~

JunHoPark93 commented on src/main/java/janggi/piece/Horse.java:
```diff
@@ -0,0 +1,57 @@
+        return board.isPresentSameTeam(team, end);
+    }
+
+    private boolean cannotMoveThrough(final Position start, final Position end, final Board board) {
```
Chariot등과 moveThrough로직이 겹치는거같은데 개선이 가능할까요?

yesjuhee replied:
d978451

혹시 수정된 지금 구조에서도 개선이 필요할까요?
예를 들면 `Chariot`, `Elephant`, `Horse`의 구현 내용이 동일한데 이 부분을 메서드로 빼서 상위 클래스에 위치시켜야 한다고 생각하시나요?
제가 봤을 때는 당장의 재사용성을 기를 수 있지만 유연성을 떨어트릴 것 같습니다!

JunHoPark93 replied:
지금 구조정도면 괜찮습니다.
상위클래스에 위치시키는건 적절하지 않아보이네요!

JunHoPark93 commented on src/main/java/janggi/piece/Chariot.java:
```diff
@@ -0,0 +1,57 @@
+        super(team, movingRules);
+    }
+
+    public static Chariot of(final Team team) {
```
movingRules를 몰라도 생성이 가능한 구조로 만들어주셨네요. 개인적으론 객체를 생성하는 입장에서 고려할 점이 적어져서 좋은것같네요.
이 부분을 의도하신걸까요?

yesjuhee replied:
오 네 맞습니다! 객체가 고정된 이동 방법을 가지고 있고, 이를 고정값으로 초기화 하기 위해 `private` 생성자로 변경하고 정적 팩터리 메서드를 사용했습니다!

JunHoPark93 commented on src/main/java/janggi/piece/Cannon.java:
```diff
@@ -0,0 +1,61 @@
+        return Type.CANNON;
+    }
+
+    private boolean cannotMoveThrough(final Position start, final Position end, final Board board) {
+        final MovingRule matchRule = findMatchRule(start, end);
+        Position route = start;
+        int count = 0;
+        for (MoveVector vector : matchRule.getVectorsWithoutLast()) {
+            route = route.add(vector);
+            if (board.isExistCannon(route)) {
+                return true;
+            }
+            if (board.isPresent(route)) {
+                count++;
+            }
+        }
+        return count != 1;
+    }
```
경로에 장애물이 2개 이상 나타나면 더 이상 확인할 필요 없이 즉시 true를 반환할 수 있을거같은데 최적화를 해보면 어떨까요?

yesjuhee replied:
d978451

count가 2인 경우 바로 반환하도록 수정했습니다!

JunHoPark93 commented on src/main/java/janggi/rule/MovingRules.java:
```diff
@@ -0,0 +1,32 @@
+        this.movingRules = movingRules;
+    }
+
+    public boolean cannotFindRule(final Position start, final Position end) {
+        final List<MoveVector> positionDiffs = new ArrayList<>();
```
적절하게 rule들을 감싸는 클래스를 만들어주셨네요. 확실히 응집이 되어 사용처에서 깔끔하게 사용이 되는거같네요.
역할을 부여하신만큼 테스트코드를 추가해보는건 어떨까요?

JunHoPark93 commented on src/main/java/janggi/piece/Piece.java:
```diff
@@ -0,0 +1,28 @@
+        this.movingRules = movingRules;
+    }
+
+    public abstract boolean canMove(final Position start, final Position end, final Board board);
+
```
canMove로 추상화를 잘 해주셨는데요, 이 내부의 구현을 보면
정의해주신 moveThrough와 rule을 찾고, 아군을 체크하는 로직이 단계별로 구현이 되어있습니다. 

move의 흐름은 모든 기물이 동일하고, 흐름의 상세는 기물마다 중복되는 부분들이있고 특정 기물의 특성을 갖는 동작이 있는데요, 흐름의 상세를 각 기물이 구현하고 move는 상위클래스가 갖도록 개선해보는건 어떨까요?

예시)
```
 public boolean canMove(final Position start, final Position end, final Board board) {
        MovingRule rule = movingRules.findMatchRule(end.calculateMoveVector(start));
        return canMoveThrough(start, end, board, rule) && isValidDestination(end, board);
    }

    protected abstract boolean canMoveThrough(Position start, Position end, Board board, MovingRule rule);
    protected abstract boolean isValidDestination(Position end, Board board);
```
이렇게하면 공통 기능을 좀 더 뽑아낼 수 있을것으로 보이는데 가능할지 고민해보시면 좋을것같습니다.

yesjuhee replied:
d978451

이동 경로에 대한 검증 책임을 조금 수정했습니다. `canMove()`를 호출할 때 `Board`의 `this`를 넘겨주는 것은 순환참조의 위험이 있다고 판단해서 이 부분을 개선했습니다.

시작 위치와 끝 위치에 대한 검증은 `Board::validateStartPosition`, `Board::validateEndPosition`에서 하고 중간 경로에 있는 기물들을 확인하는 역할은 `Piece::validateMove()`에서 하고 있습니다.

지금 구조에서 더 개선할 게있을까요?

yesjuhee replied:
https://github.com/woowacourse/java-janggi/pull/4#discussion_r2007482943

위의 코멘트와 같은 내용입니다!

yesjuhee replied:
아 추가로 순환참조에 대해서 리뷰어님의 생각이 궁금합니다!
지금 구조에서는 문제가 없지만 this를 호출하는 메서드로 넘겨주면 순환참조로 인해 무한 호출이 발생, 스택오버플로우가 발생할 수 있을 위험이 있다고 판단해서 this를 넘겨주지 않도록 리팩터링했습니다.
this를 사용하지 않기 때문에 `Board`의 `Map<Position, Piece>`를 통째로 넘겨야 하는 단점도 있는 것 같습니다.(게터를 쓰는거나 마찬가지인 것 같아요)

장단점이 있는 것 같은데 리뷰어님이라면 어떤 부분을 선택하실 것 같나요?

JunHoPark93 replied:
board와 강한 의존에서 조금은 벗어난거같네요~
말씀주신대로 순환참조 가능성이 있는 부분은 제거하는게 맞아보입니다.
저도 말씀하신 board보다는 Map<Position, Piece> 를 넘기는게 낫다고 생각합니다.

JunHoPark93 replied:
+그리고 움직임을 보드가 아닌 기물이 판단했을때의 장점을 뭐라고 생각하시나요?

JunHoPark93 reviewed (changes requested):
안녕하세요 노랑~
장기 미션 잘 구현해주셨습니다.
전체적으로 코드 깔끔하게 잘 만들어주셨네요.
몇 가지 코멘트 남겼는데 확인 부탁드릴게요~

yesjuhee committed:
refactor: 기물 이동시 검증 로직의 구조 변경

- 메서드 네이밍 변경
- Board에서 Piece의 메서드를 호출할 때 this를 넘겨주지 않도록 수정
- 최종 위치의 검증은 Board에서 수행
- 중간 경로에 대한 검증은 Piece에서 수행

yesjuhee commented:
리뷰 감사합니다 :)
우선 이동 로직의 검증 부분에 변화가 좀 많아서 이 부분에 대해 먼저 확인 받고, 논의를 통해 더 반영하고 싶은 내용이 있어서 리뷰 요청 보냈습니다! 부족한 테스트는 최대한 빨리 추가하겠습니다 🫡

JunHoPark93 commented on src/main/java/janggi/piece/Guard.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    @Override
+    public void validateMove(final Position start, final Position end, final Map<Position, Piece> board) {
```
비어져있는건 의도하신걸까요?

JunHoPark93 commented on src/main/java/janggi/piece/Horse.java:
```diff
@@ -0,0 +1,37 @@
+    public void validateMove(final Position start, final Position end, final Map<Position, Piece> board) {
+        final MovingRule matchRule = movingRules.findMatchRule(start, end);
+        Position route = start;
+        for (MoveVector vector : matchRule.getVectorsWithoutLast()) {
```
코드가 좀 더 간결해졌네요 👍

JunHoPark93 reviewed (approved):
안녕하세요 노랑~
리뷰 반영하느라 고생많으셨습니다.
부족한 테스트는 다음 단계에 이어서 반영해주시면 될 것 같습니다.
다음 단계에서 뵐게요~
